### PR TITLE
Delegation support for LibCephFS (v3)

### PR DESCRIPTION
**Pull request by jtlayton**
Github wouldn't allow me to reopen the earlier PR, so here's a new one. This one is more or less the one I had most recently + a couple of patches that rework the delegation enablement to be more like what @gregsfortytwo suggested in this PR:

https://github.com/ceph/ceph/pull/18111

I've added that rework on top of the earlier series as a couple of SQUASH patches, to make them easier to review.

### CODE REVIEW

File: src/client/Client.cc [hunk 1]
```diff
@@ -8438,6 +8454,29 @@ int Client::_open(Inode *in, int flags, mode_t mode, Fh **fhp,
...
+        need |= CEPH_CAP_FILE_RD;
+
+      result = get_caps(in, need, want, &have, -1);
+      if (result < 0) {
+	ldout(cct, 1) << "Unable to get caps after open of inode " << *in <<
+			  " . Denying open: " <<
+			  cpp_strerror(result) << dendl;
+	in->put_open_ref(cmode);
+      } else {
+	put_cap_ref(in, need);
```
Comment (ukernel) [inline src/client/Client.cc]: put_cap_ref(in, have);

**Comment by ukernel** [comment]
LGTM

File: src/client/Client.cc (hunk 1 above)
Comment (jtlayton) [reply src/client/Client.cc]: I'm not sure that's right. Sadly, get_caps doesn't have a comment header to explain what the semantics should be, but we do have this:

`    if (!waitfor_caps && !waitfor_commit) {                                     
      if ((have & need) == need) {                                              
        int revoking = implemented & ~have;                                     
        ldout(cct, 10) << "get_caps " << *in << " have " << ccap_string(have)   
                 << " need " << ccap_string(need) << " want " << ccap_string(want)
                 << " revoking " << ccap_string(revoking)                       
                 << dendl;                                                      
        if ((revoking & want) == 0) {                                           
          *phave = need | (have & want);                                        
          in->get_cap_ref(need);                                                
          return 0;                                                             
        }                                                                       
      }                                                                         
`

It looks like we only tape cap references on "need", not what ends up in "have".

Comment (ukernel) [reply src/client/Client.cc]: Oh, you are right

**Comment by jtlayton** [comment]
Patches squashed and ready for merge now, I think.

**Comment by batrick** [comment]
Not to myself: this is waiting on changes to use `mds_session_autoclose` to evict/blacklist clients that do not release caps.

**Comment by jtlayton** [comment]
Ok, two patches here. The SQUASH patch changes the new delegation timeout handling code to gate it on mds_session_autoclose instead of mds_session_timeout. The other patch can remain independent and removes mds_cap_revoke_timeout, converting its existing users to use mds_session_timeout instead.

**Comment by batrick** [comment]
@jtlayton, there was a test failure:

/ceph/teuthology-archive/pdonnell-2017-10-26_13:42:21-fs-wip-pdonnell-testing-20171025.234245-testing-basic-smithi/1776470/teuthology.log

````
2017-10-26T15:12:05.516 INFO:teuthology.orchestra.run.smithi114:Running: 'sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 0 ceph --cluster ceph --admin-daemon /var/run/ceph/ceph-mon.a.asok config get mds_revoke_cap_timeout'
2017-10-26T15:12:05.618 INFO:teuthology.orchestra.run.smithi114.stdout:{
2017-10-26T15:12:05.618 INFO:teuthology.orchestra.run.smithi114.stdout:    "error": "error getting 'mds_revoke_cap_timeout': (2) No such file or directory"
2017-10-26T15:12:05.618 INFO:teuthology.orchestra.run.smithi114.stdout:}
2017-10-26T15:12:05.619 INFO:tasks.cephfs.filesystem:_json_asok output: {
    "error": "error getting 'mds_revoke_cap_timeout': (2) No such file or directory"
}
````

The QA suite needs updated to update the config option.

File: src/client/Client.cc [hunk 2]
```diff
@@ -8438,6 +8454,29 @@ int Client::_open(Inode *in, int flags, mode_t mode, Fh **fhp,
     req->head.args.open.old_size = in->size;   // for O_TRUNC
     req->set_inode(in);
     result = make_request(req, perms);
+
+    /*
+     * Clients that have delegations enabled will wait on minimal caps, just
+     * to ensure that other clients holding delegations return theirs first.
+     */
+    if (deleg_timeout && result == 0) {
```
Comment (batrick) [inline src/client/Client.cc]: `deleg_timeout` may be used uninitialized here:

/ceph/teuthology-archive/pdonnell-2017-10-26_13:42:21-fs-wip-pdonnell-testing-20171025.234245-testing-basic-smithi/1776790/remote/smithi016/log/valgrind/client.0.log.gz

**Comment by jtlayton** [comment]
I can certainly fix up the QA suite (looking at it now). Is this something that users might hit on an upgrade? Do we need to consider leaving the setting in place for now, making it a no-op and have the code throw a warning if someone tries to set it to non-default value?

File: src/client/Client.cc (hunk 2 above)
Comment (jtlayton) [reply src/client/Client.cc]: Good catch. I missed the initialization of deleg_timeout. I'm testing the fix now.

**Comment by gregsfortytwo** [comment]
Okay, I don't exactly *like* this timeout situation, but as Jeff mentioned it's all hidden behind the interfaces so we can fix it up without having to change Ganesha or other users.

And as we get more in to things I remember more the nightmare of our distinction between the session timeout and autoclose variables. (IIRC there's this idea that clients can reconnect inside of their disconnect, but none of the code for that has ever existed or worked properly, so it just tends to break things?)

So I think we should merge this and queue up the timeouts for more careful work during the Mimic cycle. @jcsp, since we discussed the timeouts a bit last week, any concerns with that?

**Comment by jtlayton** [comment]
Thanks @gregsfortytwo. Does this mean that you'd rather we not take this patch just yet?

    mds: fold mds_revoke_cap_timeout into mds_session_timeout 

I think the main takeaway from all of this (for me anyway) is that ceph has quite a few different timeout settings in this area, and it's not at all well-defined how they work, particularly in conjunction with one another.

That tends to be a recipe for unintended emergent behavior in a complex system like ceph and I agree that it's something we really do need to get a handle on.

**Comment by jtlayton** [comment]
ceph-qa-suite PR is here:

https://github.com/ceph/ceph-qa-suite/pull/1313

Though if we're going to hold off on the patch that merges the two settings, we probably won't want to take that PR.

**Comment by jcsp** [comment]
I'm comfortable with merging mds_revoke_cap_timeout into mds_session_timeout, treating them as a unified "healthy clients respond this fast" timeout.  It still makes sense to have a separate timeout (mds_session_autoclose=300) for the "we take extreme action (hard eviction) against clients who haven't responded by now" threshold.

**Comment by jtlayton** [comment]
Thanks, @jcsp. Patches squashed and should now be ready for merge

**Comment by jtlayton** [comment]
Re-pushed to fix qa suite to handle the removal of mds_revoke_cap_timeout.

File: src/include/cephfs/libcephfs.h [hunk 3]
```diff
@@ -1572,6 +1572,85 @@ int ceph_ll_getlk(struct ceph_mount_info *cmount,
...
+ * in time, we mandate that any application wanting to use delegations
+ * explicitly set the timeout beforehand. Until this call is done on the
+ * mount, attempts to set a delegation will return -ETIME.
+ *
+ * Once a delegation is recalled, if it is not returned in this amount of
+ * time, the cmount will be forcibly unmounted and further access attempts
+ * will fail (usually with -ENOTCONN errors).
+ *
+ * This value is further vetted against the cap return timeout, and this call
+ * can fail with -EINVAL if the timeout value is too low. Delegations can be
```
Comment (batrick) [inline src/include/cephfs/libcephfs.h]: I think `timeout value is too long`?

File: src/client/Client.cc [hunk 4]
```diff
@@ -8438,6 +8455,29 @@ int Client::_open(Inode *in, int flags, mode_t mode, Fh **fhp,
     req->head.args.open.old_size = in->size;   // for O_TRUNC
     req->set_inode(in);
     result = make_request(req, perms);
+
+    /*
+     * Clients that have delegations enabled will wait on minimal caps, just
+     * to ensure that other clients holding delegations return theirs first.
```
Comment (batrick) [inline src/client/Client.cc]: Please add an explanation to this comment on why we need to do this for posterity.

**Comment by jtlayton** [comment]
Added squash patch with comment updates.

**Comment by jtlayton** [comment]
I went ahead and squashed the comment changes into the main patch.

**Comment by batrick** [comment]
http://pulpito.ceph.com/pdonnell-2017-11-20_17:50:49-fs-wip-pdonnell-testing-20171120.145001-testing-basic-smithi/#

**Comment by batrick** [comment]
http://pulpito.ceph.com/pdonnell-2017-11-21_04:57:39-fs-wip-pdonnell-testing-20171120.145001-testing-basic-smithi/

**Pull request merged**